qiskit_nature/problems/second_quantization/vibrational/vibrational_labels_validator.py

```diff
@@ -0,0 +1,120 @@
+# This code is part of Qiskit.
+#
+# (C) Copyright IBM 2021.
+#
+# This code is licensed under the Apache License, Version 2.0. You may
+# obtain a copy of this license in the LICENSE.txt file in the root directory
+# of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+#
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+"""Validator of Vibrational Labels."""
+import re
+from typing import List, Tuple, Union
+
+
+def validate_vibrational_labels(vibrational_labels: List[Tuple[str, float]], num_modes: int,
+                                num_modals: Union[int, List[int]]):
+    """
+        Validates vibrational labels in the following aspects:
+            - vibrational labels stored in a correct data structure,
+            - labels for each coefficient conform with a regular expression,
+            - indices of operators in each label are correct and ordered correctly:
+                * indices for modes and modals do not exceed declared ranges,
+                * there are no duplicated operators for each coefficient,
+                * operators in each label are sorted in the decreasing order of modes and modals,
+                if both are equal then '+' comes before '-',
+                * for each `+` operator in each label, there is a corresponding '-' operator
+                acting on the same mode.
+
+        Args:
+            vibrational_labels: list of vibrational labels with coefficients.
+            num_modes: the number of modes.
+            num_modals: the number of modals.
+
+        Raises:
+            ValueError: if invalid vibrational labels provided.
+        """
+    if isinstance(num_modals, int):
+        num_modals = [num_modals] * num_modes
+
+    _validate_data_type(vibrational_labels)
+    _validate_regex(vibrational_labels)
+    _validate_indices(vibrational_labels, num_modes, num_modals)
+
+
+def _validate_data_type(vibrational_labels: List[Tuple[str, float]]):
+    if not isinstance(vibrational_labels, list):
+        raise ValueError("Invalid data type.")
+
+
+def _validate_regex(vibrational_labels: List[Tuple[str, float]]):
+    valid_vibr_label_pattern = re.compile(r"^([\+\-]_\d+\*\d+\s)*[\+\-]_\d+\*\d+(?!\s)$|^[\+\-]+$")
+    invalid_labels = [label for label, _ in vibrational_labels if
+                      not valid_vibr_label_pattern.match(label)]
+    if invalid_labels:
+        raise ValueError(f"Invalid labels: {invalid_labels}")
+
+
+def _validate_indices(vibrational_labels: List[Tuple[str, float]], num_modes: int,
+                      num_modals: List[int]):
+    for labels, _ in vibrational_labels:
+        coeff_labels_split = labels.split(" ")
+        check_list = [0] * num_modes
```

Comment:
Either a better name or a comment explaining what this variable is used for would be great :+1: